src/Storages/ExternalStream/Kafka/KafkaSource.cpp

```diff
@@ -90,6 +92,19 @@ Chunk KafkaSource::generate()
     if (isCancelled())
         return {};
 
+    /// For non-streaming queries
+    if (high_watermark != Kafka::NO_WARTERMARK)
```

Comment:
do we need this check ? I think checking `if (offset == high_watermark)` is good to go

Reply:
Yes, a special case is `offset` and `high_watermark` both equal to `-1`. I think we should better check `high_watermark != Kafka::NO_WARTERMARK` to make sure that we won't run into some conner cases.

Reply:
NO_WARTERMARK can be -99999 to simplify logic since `-1` is used already

Reply:
We can't remove the check of `if (offset == high_watermark)` no matter what value `Kafka::NO_WATERMARK` is. Because we need to check `lastProcessedSN() >= high_watermark - 1`.

In the latest commit, I have introduced a new flag `reached_the_end`. And now the check is simpler, and more efficient.